### DISCUSSION TITLE
return statement with a block of code

### DISCUSSION THREAD

**Discussion by DarkSky7** [uNetworking/uWebSockets#1429] [General]
Line 93 of uWebSockets-master/src/HttpRouter.h's return value has a block of code attached.  I'm not familiar with this C++ construct.  Similarly, HttpRouter.cpp has numerous r.add() calls where a block of code is included within the function call.  

Can someone please refer me to C++ documentation that explains the syntax and utility of such statements?

**Comment by e3dio** [comment]
It's a function being passed to the [std::upper_bound](https://en.cppreference.com/w/cpp/algorithm/upper_bound) function. I don't know C++ but it's pretty easy to see

**Comment by ghost** [comment]
It's simpler than you think - the reason why it looks weird is because there are a whole bunch of stuff going on. It still follows the basics of C, you just need to follow it in steps until you see what it does.

**Reply by DarkSky7** [reply]
Thanks, yes, the function passed to std::upperbound( is:
[parent, this](auto &a, auto &b)
{   if ( a->isHighPriority != b->isHighPriority )
        return a->isHighPriority;
     return b->name.length() && (parent != &root) && (lexicalOrder(b->name) < lexicalOrder(a->name));
}
which makes...
   std::move(newNode)
the last parameter of parent->children.emplace(), which is itself a 'handler' that runs get()

Confusing...

**Comment by ghost** [comment]
It's a sorted insert https://en.cppreference.com/w/cpp/container/vector/emplace
